### DISCUSSION TITLE
HDR to SDR tonemapping profile/setup (gpu-next)

### DISCUSSION THREAD

**Discussion by geextahslex** [mpv-player/mpv#14757] [Show and tell]
Hi, I worked long on that HDR profile. Actually it was the reason how I got to mpv in the first place (since v.36) and finally I am satisfied. I made many comparisons in dark, bright, midtone scenes. Also over multiple productions, not only one sample. In very saturated once also. I wouldn't call it perfect as there are always compromises with tonemapping, but it is better than the default mpv settings. You can set it up and enjoy your content.  
The strong side of this project is the customizability, you can go to every shader and tune it how you want, turn it on or off and see what suits you best. It's not limited like the most build in once, where you have one setting for nits and the other one for gammut-mapping.
It is a combination of [mpv](https://github.com/mpv-player/mpv), [hdr-toys](https://github.com/natural-harmonia-gropius/hdr-toys) and a few custom-made shaders/script 

honorable mentions: @natural-harmonia-gropius from [hdr-toys](https://github.com/natural-harmonia-gropius/hdr-toys), @butterw from [bShaders](https://github.com/butterw/bShaders), folk from discord and reddit.

### Update (11.09.26): New Test Build (beta)
At the moment there is a WIP testing-build for new HDR tonemapping improvements by @kasper93
https://github.com/mpv-player/mpv/issues/15340#issuecomment-5231983486
This build is not "complete" that means that some feature like audio filters (rubberband, loudnorm) are missing, also other stuff doesnt behave "normally" so don't use it for day to day use but more for pure HDR content. I have run this build for a while and it gives great image quality, stability and performance.
There is no information on when these changes will get pushed to the main build.

This is pure mpv with config (below), no glsl shaders or other scripts
https://github.com/Dkfc123/mpv/actions/runs/34814005841/artifacts/10335924422 (newer windows version with fixes)
Download https://nightly.link/kasper93/mpv/actions/runs/31319028305
config:

```
[HDR]
profile-cond=get("video-params/primaries") == "bt.2020" and get("video-params/gamma") == "pq"
profile-restore=copy
hwdec=no
vo=gpu-next
dither-depth=8
dither=fruit

#Set saturation to your taste
saturation=-20

#PERFORMANCE (for low spec machines, better performance)
profile=fast
deband=no

sub-color='#797747'
target-colorspace-hint=no
vf=format:dolbyvision=no:hdr10plus=no
tone-mapping=st2094-10
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect=no
target-trc=bt.1886

hdr-peak-percentile=99.9
hdr-peak-decay-rate=200
hdr-scene-threshold-low=2.0
hdr-scene-threshold-high=6.0

libplacebo-opts=film_strength=0.5
hdr-reference-white=100
```

### Update (17.07.26):
This is an additional shader that boosts brightness automatically, it's meant to help with very dark series like "Silo" or "House of the Dragon" It's a variation of the [shader I made here ](https://github.com/mpv-player/mpv/discussions/18119)
You can control it on the fly with your keybind (in this example it's `W` and `S`)
A lower EFFECT_START value will start boosting at a darker condition, a higher value will start boosting at a brighter condition.
```
SHARP change-list glsl-shaders toggle "~~/shaders/auto_exposure4x.glsl"
w cycle-values glsl-shader-opts EFFECT_START=0.5 EFFECT_START=0.55 EFFECT_START=0.6 EFFECT_START=0.65 EFFECT_START=0.7 EFFECT_START=0.75 EFFECT_START=0.8 EFFECT_START=0.85 EFFECT_START=0.9 EFFECT_START=0.95 EFFECT_START=1
s cycle-values !reverse glsl-shader-opts EFFECT_START=0.5 EFFECT_START=0.55 EFFECT_START=0.6 EFFECT_START=0.65 EFFECT_START=0.7 EFFECT_START=0.75 EFFECT_START=0.8 EFFECT_START=0.85 EFFECT_START=0.9 EFFECT_START=0.95 EFFECT_START=1

```
[auto_exposure4x.zip](https://github.com/user-attachments/files/30469490/auto_exposure4x.zip)

On (with auto_exposure4x)
<img width="1920" height="1080" alt="on" src="https://github.com/user-attachments/assets/8f7e5c1f-686f-4c0b-9506-20d203049715" />
Off (default)
<img width="1920" height="1080" alt="off" src="https://github.com/user-attachments/assets/1da0841c-25c8-475e-ada8-3e29809ab63a" />

### Download:
[mpv.HDR.build.2026.zip](https://github.com/user-attachments/files/24982420/mpv.HDR.build.2026.zip) (new latest January 2026)
update (January 2026): improved gamma.lua script 

[mpv HDR build 2025.zip](https://github.com/user-attachments/files/21734004/mpv.HDR.build.2025.zip) (old August 2025)
update (August 2025): 
1. the new script will automatically boost gamma in dark scenes and reduce brightness in very bright scenes
2. I added `af=lavfi=[dynaudnorm=p=0.65:m=2:f=100:g=15:s=30]` to normalize audio (control peaks, better dialogue/speech), you can disable it for sound systems etc. `m=2` (max gain) boosts quiet parts and `p=0.65` limits max peak to 65%

[portable_config.zip](https://github.com/user-attachments/files/16769921/portable_config.zip) (old) (2024)

### Info: 
1. In HDR (PQ) mode the UI elements (which includes subtitles) have a different appearance. There is nothing that you can do about it. I added a different subtitle color to the profile. `sub-color='#e3cd00'` (yellow)
2. Saturation is set to -20, this is my personal preference on my display. Adjust it to your preference.

### Dolby Vision Build
[Dolby Vision.zip](https://github.com/user-attachments/files/26970342/Dolby.Vision.zip)
you can also use this profile for Dolby Vision, but you loose the dynamic `hdr-compute-peak` ability (this is the issue https://github.com/mpv-player/mpv/issues/11338#issuecomment-4262028777) 
To make this work you have to disable `vf=format:dolbyvision=no:hdr10plus=no` and delete the lua script from scripts folder.

```
[HDR]
profile-cond=get("video-params/primaries") == "bt.2020" and get("video-params/gamma") == "pq"
profile-restore=copy
hwdec=no
vo=gpu-next
dither-depth=8
dither=fruit

#Set saturation to your taste
saturation=-20

#PERFORMANCE (for low spec machines, better performance)
profile=fast
deband=no

#AUDIO (1. normalizes audio, 2. optional profile for multichannel 5.1)
af=lavfi=[dynaudnorm=p=0.50:m=2:f=100:g=15:s=30]
#af=lavfi=[pan=6c|c0=0.6*c0|c1=0.6*c1|c2=c2|c3=c3|c4=0.6*c4|c5=0.6*c5],lavfi=[dynaudnorm=p=0.50:m=2:f=100:g=15:s=30]

#SUBTITLES (fixes subtitle color in HDR mode)
sub-color='#e3cd00'

#HDR
target-colorspace-hint=no
gamut-mapping-mode=clip
vf=format:dolbyvision=no:hdr10plus=no
tone-mapping=st2094-10
target-peak=100
hdr-compute-peak=yes
allow-delayed-peak-detect=no
hdr-peak-percentile=99.9
hdr-peak-decay-rate=200
hdr-scene-threshold-low=2.0
hdr-scene-threshold-high=6.0
target-prim=bt.2020
target-trc=pq

#GLSL SHADER
glsl-shader=~~/shaders/pq_inv.glsl
glsl-shader-opts=L_hdr=1000,L_sdr=100
glsl-shader=~~/shaders/black_point_compensation.glsl
glsl-shader=~~/shaders/jedypod.glsl
glsl-shader=~~/shaders/bt1886.glsl

glsl-shader=~~/shaders/desatHDR.glsl
glsl-shader=~~/shaders/vibrance.glsl
#glsl-shader=~~/shaders/auto_exposure4x.glsl (optional)
```

difference between this and default mpv:
default mpv will desaturate dark scenes almost to black and white while my profile keeps the saturation normal
I opened issues for that here https://github.com/haasn/libplacebo/issues/304 and here https://github.com/mpv-player/mpv/issues/15340

this profile:
<img width="1920" height="1080" alt="mine" src="https://github.com/user-attachments/assets/e262edf2-584c-477b-a650-e7f73e4f68de" />

default mpv:
<img width="1920" height="1080" alt="mpvs" src="https://github.com/user-attachments/assets/b71fc8ac-356b-4e8d-abaa-6ecec73a8038" />

**Comment by Vantskruv** [comment]
Thank you! I am using settings and I am very satisfied.
I am running a HTPC with Kodi as browser and MPV as player (overriding the Kodi player).
The HTPC is connected to a RX-A8A reciever and a Sony HW65ES projector.

**Reply by geextahslex** [reply]
here is an updated version https://github.com/mpv-player/mpv/issues/15340#issuecomment-2509801760
just add to that `hdr-peak-percentile=99.8`
is much better know

**Reply by Vantskruv** [reply]
Thanks! I will check this out tonight when watching Aquaman. :)

**Reply by kasper93** [reply]
> here is an updated version [#15340 (comment)](https://github.com/mpv-player/mpv/issues/15340#issuecomment-2509801760) just add to that `hdr-peak-percentile=99.8` is much better know

You should consider `profile=high-quality`

**Reply by geextahslex** [reply]
>You should consider `profile=high-quality`

In the future I will, but not on my current potato PC. It's a wonder that HDR tonemapping even functions ^^'

**Reply by Vantskruv** [reply]
I wonder if this would give a better image-quality, if running on a HDR projector, hence ignoring the projector "HDR capabilites", as projectors is known to not work well with HDR and Dolby Vision?

I am on the razors edge of buying a 4k projector (which often do have HDR support). I am not sure if it is worth it, I am currently satisfied with the image quality from my 10 year old Sony HW65ES (I never seen a 4k projector in action).

**Reply by geextahslex** [reply]
I have no experiences with projectors, but when you consider that many HDR TV's have issues with displaying proper HDR I would think twice about that. HDR tonemapping makes a good job, when the picture is bright enough for you I would wait with any purchase. Only my opinion.

**Reply by danbezerra** [reply]
I have a JVC projector and use MPV to tone mapping HDR to SDR gamma, but configured to use my max nits output (~100 nits, to add some extra highlights) and BT.2020. So better than regular SDR.

There are some newer projectors with good HDR/DV tone mapping built in as well.

**Comment by Canowyrms** [comment]
Nice, looking forward to trying this.

Have you given any thought to setting up a repo with releases? Might be a little easier for users to receive updates 🤔

Regardless, thanks for this!

**Reply by geextahslex** [reply]
Honestly I don't know how all fo this repo/github stuff works, so it's hard to say. I'm just a casual user.
No problem, I hope it works nice for you :)

**Reply by glpc-blaze** [reply]
can these shaders and scripts work on android mpv i'm amazed by this work and works finely on windows

**Reply by geextahslex** [reply]
I don't know, I never tried it on android

**Comment by dasdasrft** [comment]
I used your preset and it looks great, but I’m wondering why you set the saturation to -20. Is it just a personal preference? It looks a bit washed out on my monitor.

**Reply by geextahslex** [reply]
Thats nice to hear :)
Yes depending on the content I switch between -10 and -25 saturation. Some productions (like sci-fi shows) are oversaturated for my taste, but yes this is only my preference.

**Comment by Headcrabed** [comment]
Btw what would it look like if using `tone-mapping=bt.2446a`? I found someone said that this would give better result among mpv's built-in tonemapping algorithms.

**Reply by geextahslex** [reply]
You would have to make screenshots from some samples with low,mid and highlight scenes and then see which one looks best to you.

**Comment by rootrascal** [comment]
This is really good. Although, it seems like the shaders are being applied to the entire window rather than just the media.

I noticed my UI controls and subtitles are greyed out now and are kinda hard to see since they aren't their regular bright white anymore. Is there a way to fix this?

**Reply by geextahslex** [reply]
No, unfortunately you can't bypass this. You can adjust your subtitles color and add this to your HDR profile. I use this yellow color 
`sub-color='#e3cd00'`

**Comment by stellarwork** [comment]
Would this work with mpv-android? Thanks.

**Reply by geextahslex** [reply]
I don't know. I only use it on Windows

**Reply by glpc-blaze** [reply]
works fine on android but some shaders like vibrance struggle

**Comment by glpc-blaze** [comment]
im seeing visible brightness fluctuations using your script in bright scenes

**Reply by natural-harmonia-gropius** [reply]
Try the latest hdr-toys, this issue should improve.

**Reply by geextahslex** [reply]
remove `peak smooth advanced gamma.lua` from scripts folder and check if this is causing the issue. If it is, then lower the `local easing_speed = 1.5`
It can also be `hdr-compute-peak=yes` which has this effect on bright first frame scenes, but there is nothing I can do about it, it's how mpv handles it.

**Reply by glpc-blaze** [reply]
indeed, im using your shaders  .latest April version i think astra shader makes brightness fluctuation 
requesting make a fix

**Reply by glpc-blaze** [reply]
@geextahslex ive done that earlier its not the script problem im using astra with it that may be causing problem as when i removed it flucation didnt happened

**Reply by geextahslex** [reply]
This script is finetuned as a package with all shaders, scripts and settings.
[Astra](https://github.com/natural-harmonia-gropius/hdr-toys/blob/master/shaders/hdr-toys/tone-mapping/astra.glsl) is its own thing that controls exposure, brightness etc. on its own. These are two different paths. You can't slap them together.
To make astra work you have to follow [hdr-toys](https://github.com/natural-harmonia-gropius/hdr-toys) [setup/config](https://github.com/natural-harmonia-gropius/hdr-toys/blob/master/hdr-toys.conf)
It will not work properly with astra as glsl from hdr-toys. This script works only with mpv native tone-mappers like `tone-mapping=st2094-10`

**Reply by glpc-blaze** [reply]
glsl-shader-opts=L_hdr=1000,L_sdr=100 what should be the value of sdr as when i set to 300{my ips's peak brightness} i get dark video

**Reply by geextahslex** [reply]
it should be the same value as `target-peak=100`

you can adjust this on the fly, put this into your config, and keybind to something
L_sdr
```
PGUP cycle-values glsl-shader-opts pq_inv/L_sdr=50 pq_inv/L_sdr=60 pq_inv/L_sdr=70 pq_inv/L_sdr=80 pq_inv/L_sdr=90 pq_inv/L_sdr=100 pq_inv/L_sdr=110 pq_inv/L_sdr=120 pq_inv/L_sdr=130 pq_inv/L_sdr=140 pq_inv/L_sdr=150 pq_inv/L_sdr=200 
PGDWN cycle-values !reverse glsl-shader-opts pq_inv/L_sdr=50 pq_inv/L_sdr=60 pq_inv/L_sdr=70 pq_inv/L_sdr=80 pq_inv/L_sdr=90 pq_inv/L_sdr=100 pq_inv/L_sdr=110 pq_inv/L_sdr=120 pq_inv/L_sdr=130 pq_inv/L_sdr=140 pq_inv/L_sdr=150 pq_inv/L_sdr=200 
```

target-peak
```
UP add target-peak 50
DOWN add target-peak -50
```

**Comment by KAGEYAM4** [comment]
I was playing `*.1080p.NF.WEB-DL.DDP5.1.Atmos.DV.HDR.H.265-teamACH` and the color looked washed out, i have SDR display. Using yours latest zip fixed it but subtitle looked washed out, as you mentioned in previous comment and hdr-toys faq.

is there no inbuilt solution for this in mpv? is this being worked on? I saw https://github.com/mpv-player/mpv/issues/15340, but i am not sure if this is related.

I tried showtime ( gnome video player ) and it looks good there but it using alot of CPU. I tried VLC but it is same here, the situation is little bit improved in vlc-git but not fully.

is this problem niche? from now on i will be careful not to get HDR/DV files.

Also, the tonemapping is quite good on https://github.com/mpv-android or https://github.com/marlboro-advance/mpvEx, can you take a look. I think its even better than what the current config is doing.

**Reply by geextahslex** [reply]
Have you tried to change the subtitle color, like I mentioned in a previous question?
No, unfortunately you can't bypass this. You can adjust your subtitles color and add this to your HDR profile. I use this yellow color 
`sub-color='#e3cd00'`

>is there no inbuilt solution for this in mpv? is this being worked on? I saw #15340, but i am not sure if this is related.

I have no idea, I am not involved in the development of mpv, honestly this doesn't bother me

>I tried showtime ( gnome video player ) and it looks good there but it using alot of CPU. I tried VLC but it is same here, the situation is little bit improved in vlc-git but not fully.

Yes this project is very light, I have an old PC thats why performance was crucial to me

>is this problem niche?

Which problem do you mean? Tone-mapping in general? I guess it is, probably most people have already an HDR display/tv

>from now on i will be careful not to get HDR/DV files.

I always try to get HDR, this gives you the best image quality. SDR can not match this, but if you have tone-mapping issues, than go for SDR.
If you are only watching on a phone, I wouldn't use HDR, files are bigger, quality on a small screen doesn't really matter and the battery will drain faster.

>Also, the tonemapping is quite good on https://github.com/mpv-android or https://github.com/marlboro-advance/mpvEx, can you take a look. I think its even better than what the current config is doing.

I have no experience with those projects and also with anything on android.

**Comment by maxim-sm** [comment]
Thank you for sharing the profile! I think it looks much better than the default.

A quick question, is there a specific reason to use `hwdec=no`? Will it be OK to use `hwdec=auto-safe` with this profile?

**Reply by geextahslex** [reply]
You welcome :)
Not really (so no specific reason), this is just how I setup my config and I know it works. You can make changes and see if it still works normally.

**Reply by maxim-sm** [reply]
Got it, thanks!
